src/nni_manager/training_service/pai/paiK8S/paiK8STrainingService.ts

```diff
@@ -0,0 +1,307 @@
+} from '../../../common/trainingService';
+import { delay, generateParamFileName,
+    getExperimentRootDir, getIPV4Address, getVersion, uniqueString, unixPathJoin } from '../../../common/utils';
+import { CONTAINER_INSTALL_NNI_SHELL_FORMAT } from '../../common/containerJobData';
+import { TrialConfigMetadataKey } from '../../common/trialConfigMetadataKey';
+import { execMkdir, validateCodeDir, execCopydir } from '../../common/util';
+import { PAI_K8S_TRIAL_COMMAND_FORMAT } from './paiK8SData';
+import { NNIPAIK8STrialConfig } from './paiK8SConfig';
+import { PAIK8SJobRestServer } from './paiK8SJobRestServer';
+import { PAITrainingService } from '../paiTrainingService';
+import { PAIClusterConfig, PAITrialJobDetail } from '../paiConfig';
+
+const yaml = require('js-yaml');
+
+/**
+ * Training Service implementation for OpenPAI (Open Platform for AI)
+ * Refer https://github.com/Microsoft/pai for more info about OpenPAI
+ */
+@component.Singleton
+class PAIK8STrainingService extends PAITrainingService {
+    protected paiTrialConfig: NNIPAIK8STrialConfig | undefined;
+
+    constructor() {
+        super();
+    }
+
+    public async setClusterMetadata(key: string, value: string): Promise<void> {
+        const deferred: Deferred<void> = new Deferred<void>();
+
+        switch (key) {
+            case TrialConfigMetadataKey.NNI_MANAGER_IP:
+                this.nniManagerIpConfig = <NNIManagerIpConfig>JSON.parse(value);
+                deferred.resolve();
+                break;
+
+            case TrialConfigMetadataKey.PAI_CLUSTER_CONFIG:
+                this.paiJobRestServer = component.get(PAIK8SJobRestServer);
+                this.paiClusterConfig = <PAIClusterConfig>JSON.parse(value);
+
+                if(this.paiClusterConfig.passWord) {
+                    // Get PAI authentication token
+                    await this.updatePaiToken();
+                } else if(this.paiClusterConfig.token) {
+                    this.paiToken = this.paiClusterConfig.token;
+                } else {
+                    deferred.reject(new Error('pai cluster config format error, please set password or token!'));
+                }
+
+                deferred.resolve();
+                break;
+
+            case TrialConfigMetadataKey.TRIAL_CONFIG:
+                if (this.paiClusterConfig === undefined) {
+                    this.log.error('pai cluster config is not initialized');
+                    deferred.reject(new Error('pai cluster config is not initialized'));
+                    break;
+                }
+                this.paiTrialConfig = <NNIPAIK8STrialConfig>JSON.parse(value);
+
+                // Validate to make sure codeDir doesn't have too many files
+                try {
+                    await validateCodeDir(this.paiTrialConfig.codeDir);
+                } catch (error) {
+                    this.log.error(error);
+                    deferred.reject(new Error(error));
```

Comment:
this error should have already been logged by rest server,   no need to catch and re-throw, please check it.

Reply:
fixed.